descriptor/qsarmolecular/src/main/java/org/openscience/cdk/qsar/descriptors/molecular/JPlogPDescriptor.java

```diff
@@ -0,0 +1,874 @@
+ */
+public class JPlogPDescriptor extends AbstractMolecularDescriptor implements IMolecularDescriptor {
+
+	private static final String[] NAMES = { "JPLogP" };
+	private boolean addImplicitH = true;
+	JPlogPCalculator jplogp = null;
+
+	/**
+	 * Default constructor which will setup the required coefficients to enable
+	 * a prediction
+	 */
+	public JPlogPDescriptor() 
+	{
+		jplogp = new JPlogPCalculator();
+	}
+
+	@Override
+	public DescriptorSpecification getSpecification() {
+		return new DescriptorSpecification("JPlogP developed at Lhasa Limited www.lhasalimited.org",
+				this.getClass().getName(), "Jeffrey Plante - Lhasa Limited");
+	}
+
+	@Override
+	public String[] getParameterNames() {
+		String[] params = new String[1];
+		params[0] = "addImplicitH";
+		return params;
+	}
+
+	@Override
+	public Object getParameterType(String name) {
+		if ("addImplicitH".equals(name))
+			return true;
+		return null;
+	}
+
+	@Override
+	public void setParameters(Object[] params) throws CDKException {
+		if (params.length != 1) {
+			throw new CDKException("JPLogPDescriptor expects one parameter");
+		}
+		if (!(params[0] instanceof Boolean)) {
+			throw new CDKException("The parameter must be of type Boolean");
+		}
+		addImplicitH = (Boolean) params[0];
+	}
+
+	@Override
+	public Object[] getParameters() {
+		Object[] params = new Object[1];
+		params[0] = addImplicitH;
+		return params;
+	}
+
+	@Override
+	public String[] getDescriptorNames() {
+		return NAMES;
+	}
+
+	@Override
+	public IDescriptorResult getDescriptorResultType() {
+		return new DoubleResult(0.0);
+	}
+
+	private DescriptorValue getDummyDescriptorValue(Exception e) {
+		return new DescriptorValue(getSpecification(), getParameterNames(), getParameters(),
+				new DoubleResult(Double.NaN), getDescriptorNames(), e);
+	}
+
+	@Override
+	public DescriptorValue calculate(IAtomContainer container) {
+		IAtomContainer struct;
+		try {
+			struct = (IAtomContainer) container.clone();
+			AtomContainerManipulator.percieveAtomTypesAndConfigureUnsetProperties(struct);
+			CDKHydrogenAdder hAdder = CDKHydrogenAdder.getInstance(struct.getBuilder());
+			hAdder.addImplicitHydrogens(struct);
+			AtomContainerManipulator.convertImplicitToExplicitHydrogens(struct);
+			Aromaticity.cdkLegacy().apply(struct);
+		} catch (CloneNotSupportedException e) {
+			return getDummyDescriptorValue(e);
+		} catch (CDKException e) {
+			return getDummyDescriptorValue(e);
+		}
+
+
+		return new DescriptorValue(getSpecification(), getParameterNames(), getParameters(), new DoubleResult(jplogp.calcLogP(struct)),
+				getDescriptorNames());
+	}
+
+
+
+	/**
+	 * The class that calculated the logP according to the JPlogP method described in:
+	 * Journal of Cheminformatics 2018 10:61 https://doi.org/10.1186/s13321-018-0316-5
+	 * 
+	 * This is lower level access and should normally be obtained through the descriptor above.
+	 * 
+	 * @author Jeffrey
+	 *
+	 */
+	protected class JPlogPCalculator
```

Comment:
Extracted the JPlogP specifics into this inner class

Reply:
OK. I rather see this as external class, but that can be done later.